File: backend/app/services/filter_engine.py
"""Python mirror of src/utils/filterEngine.ts so both sides filter identically."""

from datetime import datetime
from typing import Any

from app.schemas.filter import FilterGroup, FilterRule
from app.services.type_detection import is_blank, to_boolean, to_date, to_number

OPERATOR_LABELS = {
    "eq": "is equal to",
    "neq": "is not equal to",
    "contains": "contains",
    "not_contains": "does not contain",
    "starts_with": "starts with",
    "ends_with": "ends with",
    "gt": "is greater than",
    "gte": "is greater than or equal to",
    "lt": "is less than",
    "lte": "is less than or equal to",
    "between": "is between",
    "before": "is before",
    "after": "is after",
    "on": "is on",
    "is_true": "is Yes",
    "is_false": "is No",
    "is_empty": "is empty",
    "is_not_empty": "is not empty",
    "in": "is one of",
}

NO_VALUE_OPERATORS = {"is_true", "is_false", "is_empty", "is_not_empty"}


def _text(value: Any) -> str:
    return ("" if value is None else str(value)).strip().lower()


def is_rule_complete(rule: FilterRule) -> bool:
    if not rule.column:
        return False
    if rule.operator in NO_VALUE_OPERATORS:
        return True
    if rule.operator == "between":
        return bool(rule.value.strip()) and bool(rule.value2.strip())
    return bool(rule.value.strip())


def evaluate_rule(row: dict[str, Any], rule: FilterRule) -> bool:
    raw = row.get(rule.column)
    target = rule.value.strip()

    if rule.operator == "is_empty":
        return is_blank(raw)
    if rule.operator == "is_not_empty":
        return not is_blank(raw)
    if rule.operator == "is_true":
        return to_boolean(raw) is True
    if rule.operator == "is_false":
        return to_boolean(raw) is False

    if is_blank(raw):
        return False

    number_value = to_number(raw)
    number_target = to_number(target)
    numeric = number_value is not None and number_target is not None

    if rule.operator == "eq":
        return number_value == number_target if numeric else _text(raw) == _text(target)
    if rule.operator == "neq":
        return number_value != number_target if numeric else _text(raw) != _text(target)
    if rule.operator == "contains":
        return _text(target) in _text(raw)
    if rule.operator == "not_contains":
        return _text(target) not in _text(raw)
    if rule.operator == "starts_with":
        return _text(raw).startswith(_text(target))
    if rule.operator == "ends_with":
        return _text(raw).endswith(_text(target))
    if rule.operator == "in":
        options = [part.strip().lower() for part in target.split(",") if part.strip()]
        return _text(raw) in options
    if rule.operator == "gt":
        return numeric and number_value > number_target
    if rule.operator == "gte":
        return numeric and number_value >= number_target
    if rule.operator == "lt":
        return numeric and number_value < number_target
    if rule.operator == "lte":
        return numeric and number_value <= number_target
    if rule.operator == "between":
        second = to_number(rule.value2)
        if numeric and second is not None:
            low, high = sorted([number_target, second])
            return low <= number_value <= high
        date_value, start, end = to_date(raw), to_date(target), to_date(rule.value2)
        if date_value and start and end:
            low_date, high_date = sorted([start, end])
            return low_date <= date_value <= high_date
        return False
    if rule.operator in {"before", "after", "on"}:
        date_value = to_date(raw)
        other = to_date(target)
        if not date_value or not other:
            return False
        if rule.operator == "before":
            return date_value < other
        if rule.operator == "after":
            return date_value > other
        return date_value.date() == other.date()
    return False


def apply_filters(rows: list[dict[str, Any]], group: FilterGroup) -> list[dict[str, Any]]:
    active = [rule for rule in group.rules if is_rule_complete(rule)]
    if not active:
        return rows
    combine = all if group.combinator == "AND" else any
    return [row for row in rows if combine(evaluate_rule(row, rule) for rule in active)]


def describe_group(group: FilterGroup) -> str:
    parts: list[str] = []
    for rule in group.rules:
        if not is_rule_complete(rule):
            continue
        base = f"{rule.column} {OPERATOR_LABELS[rule.operator]}"
        if rule.operator in NO_VALUE_OPERATORS:
            parts.append(base)
        elif rule.operator == "between":
            parts.append(f"{base} {rule.value} and {rule.value2}")
        else:
            parts.append(f"{base} {rule.value}")
    return f" {group.combinator} ".join(parts) if parts else "All students"


def utc_now() -> datetime:
    from datetime import timezone

    return datetime.now(timezone.utc)

File: src/utils/filterEngine.ts
import type { ColumnMeta, DataRow } from "@/types/dataset";
import {
  NO_VALUE_OPERATORS,
  OPERATOR_LABELS,
  type FilterGroup,
  type FilterRule,
} from "@/types/filter";
import { isBlank, toBoolean, toDate, toNumber } from "@/utils/dataTypeDetection";

/** A rule is usable only when it has a column and (if needed) a value. */
export function isRuleComplete(rule: FilterRule): boolean {
  if (!rule.column) return false;
  if (NO_VALUE_OPERATORS.includes(rule.operator)) return true;
  if (rule.operator === "between") return rule.value.trim() !== "" && rule.value2.trim() !== "";
  return rule.value.trim() !== "";
}

function text(value: unknown): string {
  return String(value ?? "").trim().toLowerCase();
}

function evaluateRule(row: DataRow, rule: FilterRule): boolean {
  const raw = row[rule.column];
  const target = rule.value.trim();

  switch (rule.operator) {
    case "is_empty":
      return isBlank(raw);
    case "is_not_empty":
      return !isBlank(raw);
    case "is_true":
      return toBoolean(raw) === true;
    case "is_false":
      return toBoolean(raw) === false;
    default:
      break;
  }

  if (isBlank(raw)) return false;

  const numberValue = toNumber(raw);
  const numberTarget = toNumber(target);
  const numeric = numberValue !== null && numberTarget !== null;

  switch (rule.operator) {
    case "eq":
      return numeric ? numberValue === numberTarget : text(raw) === text(target);
    case "neq":
      return numeric ? numberValue !== numberTarget : text(raw) !== text(target);
    case "contains":
      return text(raw).includes(text(target));
    case "not_contains":
      return !text(raw).includes(text(target));
    case "starts_with":
      return text(raw).startsWith(text(target));
    case "ends_with":
      return text(raw).endsWith(text(target));
    case "in":
      return target
        .split(",")
        .map((part) => part.trim().toLowerCase())
        .filter(Boolean)
        .includes(text(raw));
    case "gt":
      return numeric && numberValue > numberTarget;
    case "gte":
      return numeric && numberValue >= numberTarget;
    case "lt":
      return numeric && numberValue < numberTarget;
    case "lte":
      return numeric && numberValue <= numberTarget;
    case "between": {
      const second = toNumber(rule.value2);
      if (numeric && second !== null) {
        const low = Math.min(numberTarget, second);
        const high = Math.max(numberTarget, second);
        return numberValue >= low && numberValue <= high;
      }
      const dateValue = toDate(raw);
      const from = toDate(target);
      const to = toDate(rule.value2);
      if (dateValue && from && to) {
        const low = Math.min(from.getTime(), to.getTime());
        const high = Math.max(from.getTime(), to.getTime());
        return dateValue.getTime() >= low && dateValue.getTime() <= high;
      }
      return false;
    }
    case "before": {
      const dateValue = toDate(raw);
      const other = toDate(target);
      return !!dateValue && !!other && dateValue.getTime() < other.getTime();
    }
    case "after": {
      const dateValue = toDate(raw);
      const other = toDate(target);
      return !!dateValue && !!other && dateValue.getTime() > other.getTime();
    }
    case "on": {
      const dateValue = toDate(raw);
      const other = toDate(target);
      return (
        !!dateValue && !!other && dateValue.toDateString() === other.toDateString()
      );
    }
    default:
      return false;
  }
}

export function applyFilters(rows: DataRow[], group: FilterGroup): DataRow[] {
  const active = group.rules.filter(isRuleComplete);
  if (active.length === 0) return rows;

  return rows.filter((row) => {
    const results = active.map((rule) => evaluateRule(row, rule));
    return group.combinator === "AND" ? results.every(Boolean) : results.some(Boolean);
  });
}

/** Plain-English description a teacher can read back, e.g. "Marks is greater than 40 AND ...". */
export function describeGroup(group: FilterGroup, columns: ColumnMeta[]): string {
  const label = (key: string) => columns.find((column) => column.key === key)?.label ?? key;
  const parts = group.rules.filter(isRuleComplete).map((rule) => {
    const base = `${label(rule.column)} ${OPERATOR_LABELS[rule.operator]}`;
    if (NO_VALUE_OPERATORS.includes(rule.operator)) return base;
    if (rule.operator === "between") return `${base} ${rule.value} and ${rule.value2}`;
    return `${base} ${rule.value}`;
  });
  return parts.length === 0 ? "All students" : parts.join(` ${group.combinator} `);
}

export function toCsv(columns: ColumnMeta[], rows: DataRow[]): string {
  const escape = (value: unknown) => {
    const cell = value === null || value === undefined ? "" : String(value);
    return /[",\n]/.test(cell) ? `"${cell.replace(/"/g, '""')}"` : cell;
  };
  const header = columns.map((column) => escape(column.label)).join(",");
  const body = rows.map((row) => columns.map((column) => escape(row[column.key])).join(","));
  return [header, ...body].join("\n");
}

export function downloadCsv(filename: string, csv: string) {
  const blob = new Blob([csv], { type: "text/csv;charset=utf-8;" });
  const url = URL.createObjectURL(blob);
  const anchor = document.createElement("a");
  anchor.href = url;
  anchor.download = filename.endsWith(".csv") ? filename : `${filename}.csv`;
  anchor.click();
  URL.revokeObjectURL(url);
}
